### PR TITLE
rbx_binary implement Color3uint8

### PR DESCRIPTION
**Pull request by kennethloeffler**
As part of this PR I have made `serializer.rs` [use canonical property names when fetching property values to write, rather than the serialized names](https://github.com/rojo-rbx/rbx-dom/blob/1b2cfa7efef6556504fc9a732f13cbcad3615523/rbx_binary/src/serializer.rs#L446). This became an issue because `deserializer.rs` uses canonical names as keys for property maps. This caused the serializer to erroneously write default prop values since `instance.properties.get("Color3uint8")` results in a `None`. I alternatively could have made `deserializer.rs` use `"Color3uint8"` as the key, but that seems smelly to me.

I have also updated the test files submodule.

### CODE REVIEW

File: rbx_binary/src/deserializer.rs [hunk 1]
```diff
@@ -856,7 +856,45 @@ impl<R: Read> BinaryDeserializer<R> {
...
+                    for _ in 0..len {
+                        r.push(chunk.read_u8().unwrap());
+                    }
+
+                    for _ in 0..len {
+                        g.push(chunk.read_u8().unwrap());
+                    }
+
+                    for _ in 0..len {
+                        b.push(chunk.read_u8().unwrap());
```
Comment (Kampfkarren) [inline rbx_binary/src/deserializer.rs:875]: Are you sure you want to be using `unwrap()` here instead of `?`?

File: rbx_binary/src/text_deserializer.rs [hunk 2]
```diff
@@ -401,6 +402,32 @@ impl DecodedValues {
...
+                for _ in 0..prop_count {
+                    r.push(reader.read_u8().unwrap());
+                }
+
+                for _ in 0..prop_count {
+                    g.push(reader.read_u8().unwrap());
+                }
+
+                for _ in 0..prop_count {
+                    b.push(reader.read_u8().unwrap());
```
Comment (Kampfkarren) [inline rbx_binary/src/text_deserializer.rs:419]: here too

Comment (Dekkonot) [reply rbx_binary/src/text_deserializer.rs:419]: `read_u8` returns a Result whereas the parent function here returns an Option. If you look at the file, all of the other read functions use `unwrap` too.

File: rbx_binary/src/deserializer.rs [hunk 3]
```diff
@@ -856,7 +856,45 @@ impl<R: Read> BinaryDeserializer<R> {
...
+                    }
```
Comment (LPGhatguy) [inline rbx_binary/src/deserializer.rs:876]: Instead of writing these loops by hand, we can use Rust's [Read::read_exact](https://doc.rust-lang.org/stable/std/io/trait.Read.html#method.read_exact).

```rust
let mut r = vec![0; len];
chunk.read_exact(r.as_mut_slice())?;
```

...for each color.

File: rbx_binary/src/serializer.rs [hunk 4]
```diff
@@ -696,6 +704,33 @@ impl<'a, W: Write> BinarySerializer<'a, W> {
...
+                        let mut r = Vec::with_capacity(values.len());
+                        let mut g = Vec::with_capacity(values.len());
+                        let mut b = Vec::with_capacity(values.len());
+
+                        for (i, rbx_value) in values {
+                            if let Variant::Color3uint8(value) = rbx_value.as_ref() {
+                                r.push(value.r);
+                                g.push(value.g);
+                                b.push(value.b);
+                            } else {
```
Comment (LPGhatguy) [inline rbx_binary/src/serializer.rs:717]: We should probably also support `Color3` values here, since they're often trivially converted. If the `Color3`'s range does not fit within a `Color3uint8`, it should be clamped.

File: rbx_binary/src/serializer.rs [hunk 5]
```diff
@@ -696,6 +704,33 @@ impl<'a, W: Write> BinarySerializer<'a, W> {
...
+                            chunk.write_u8(component)?;
+                        }
+
+                        for component in g {
+                            chunk.write_u8(component)?;
+                        }
+
+                        for component in b {
+                            chunk.write_u8(component)?;
+                        }
```
Comment (LPGhatguy) [inline rbx_binary/src/serializer.rs:732]: Similar to the write examples, we have a method for this from the Rust standard library, [`Write::write_all`](https://doc.rust-lang.org/stable/std/io/trait.Write.html#method.write_all).

We should write something like...

```rust
chunk.write_all(r.as_slice())?;
```

...for each color.

File: rbx_binary/src/serializer.rs [hunk 6]
```diff
@@ -443,7 +443,15 @@ impl<'a, W: Write> BinarySerializer<'a, W> {
...
                                 .properties
-                                .get(prop_name)
+                                .get(
+                                    if let Some(descriptor) =
+                                        find_canonical_property_descriptor(type_name, prop_name)
+                                    {
+                                        &*descriptor.name
+                                    } else {
+                                        prop_name
+                                    },
```
Comment (LPGhatguy) [inline rbx_binary/src/serializer.rs:453]: This is a much too complicated expression to slap inline here. I need to think over the implications of this as well.

File: README.md [hunk 7]
```diff
@@ -66,10 +66,10 @@ Roblox Lua implementation of DOM APIs, allowing Instance reflection from inside
 | BrickColor         | `Part.BrickColor`               | ✔ | ✔ | ✔ | ❌ |
 | CFrame             | `Camera.CFrame`                 | ✔ | ✔ | ✔ | ➖ |
 | Color3             | `Lighting.Ambient`              | ✔ | ✔ | ✔ | ✔ |
-| Color3uint8        | `N/A`                           | ✔ | ✔ | ✔ | ❌ |
+| Color3uint8        | `N/A`                           | ✔ | ✔ | ✔ | ✔ |
 | ColorSequence      | `Beam.Color`                    | ✔ | ✔ | ✔ | ❌ |
 | Content            | `Decal.Texture`                 | ✔ | ✔ | ✔ | ✔ |
-| Enum               | `Part.Shape`                    | ✔ | ✔ | ✔ | ❌ |
+| Enum               | `Part.Shape`                    | ✔ | ✔ | ✔ | ✔ |
```
Comment (LPGhatguy) [inline README.md:72]: This is an unrelated change that I've already landed in master. I'd prefer if changes like this were not included in feature PRs.

**Review by LPGhatguy — CHANGES_REQUESTED** [review]

File: rbx_binary/src/serializer.rs (hunk 6 above)
Comment (LPGhatguy) [reply rbx_binary/src/serializer.rs:453]: This is also a problem because we're redundantly looking up the same property descriptor in the database for every instance. This process is expensive and can take a long time.

Comment (LPGhatguy) [reply rbx_binary/src/serializer.rs:453]: I am working on what I think is the correct solution to this problem. We should avoid introducing a fix here, especially as part of a regular property feature implementation.

Comment (kennethloeffler) [reply rbx_binary/src/serializer.rs:453]: Should I have `deserializer.rs` use `Color3uint8` as the key for the property map instead of the canonical name (as described in the original PR comment)? Or perhaps check explicitly in `serializer.rs` (similar to `Name`) and do `instance.properties.get("Color")`? I'm grepping through the lua database and it looks like this type is exclusively used by `BasePart` and friends, so either way seems like it would work. Your call.

**Comment by kennethloeffler** [comment]
Leaving that problem alone, I've made the other changes.

File: rbx_binary/src/serializer.rs (hunk 6 above)
Comment (LPGhatguy) [reply rbx_binary/src/serializer.rs:453]: We should do neither here, the change I'm working on should solve this problem with no hard-coded special cases. `Color` is not quite the same level of specialness as `Name` 😄

File: rbx_binary/src/tests/serializer.rs [hunk 8]
```diff
@@ -124,3 +124,30 @@ fn logical_properties_basepart_size() {
...
+fn color3_to_color3uint8() {
+    let tree = WeakDom::new(
+        InstanceBuilder::new("Folder")
+            .with_child(
+                InstanceBuilder::new("Part")
+                    .with_property("Color3uint8", Color3::new(-0.25, 0.5, 1.2)),
+            )
+            .with_child(
+                InstanceBuilder::new("Part")
+                    .with_property("Color", Color3::new(0.1111, 0.3333, 0.9999)),
```
Comment (LPGhatguy) [inline rbx_binary/src/tests/serializer.rs:141]: Can we change this property to be `Color3uint8` with a `Color3uint8` value? As is, this is redundant with the third instance. Testing all of the valid combinations here is a good idea I think.

**Review by LPGhatguy — APPROVED** [review]
